seastar/src/file.rs

```diff
@@ -0,0 +1,433 @@
+
+        unsafe fn read_dma(
+            file: &UniquePtr<file_t>,
+            buffer: *mut u8,
+            size: u64,
+            pos: u64,
+        ) -> IntFuture;
+
+        unsafe fn write_dma(
+            file: &UniquePtr<file_t>,
+            buffer: *mut u8,
+            size: u64,
+            pos: u64,
+        ) -> IntFuture;
+
+        fn flush(file: &UniquePtr<file_t>) -> VoidFuture;
+
+        fn close(file: &UniquePtr<file_t>) -> VoidFuture;
+
+        fn size(file: &UniquePtr<file_t>) -> IntFuture;
+    }
+}
+
+const ALIGN: usize = 512;
+const CHUNK_SIZE: usize = 4096;
+
+// A buffer that stores/receives data for I/O operations.
+// Its contents are aligned in memory up to 512 bytes.
+// read_dma and write_dma require memory to be aligned.
+pub struct DmaBuffer {
+    buffer: *mut u8,
+    size: usize,
+}
+
+impl Deref for DmaBuffer {
+    type Target = [u8];
+
+    fn deref(&self) -> &Self::Target {
+        self.as_slice()
+    }
+}
+
+impl Drop for DmaBuffer {
+    fn drop(&mut self) {
+        let layout = Layout::from_size_align(self.size, ALIGN).unwrap();
+        unsafe {
+            alloc::dealloc(self.buffer, layout);
+        }
+    }
+}
+
+impl Index<usize> for DmaBuffer {
+    type Output = u8;
+
+    fn index(&self, idx: usize) -> &Self::Output {
+        self.as_slice().index(idx)
+    }
+}
+
+impl IndexMut<usize> for DmaBuffer {
+    fn index_mut(&mut self, idx: usize) -> &mut Self::Output {
+        self.as_mut_slice().index_mut(idx)
+    }
+}
+
+impl DmaBuffer {
+    pub fn from_vec(bytes: Vec<u8>) -> Self {
+        let size = bytes.len();
+        assert!(size % CHUNK_SIZE == 0);
+        let layout = Layout::from_size_align(size, ALIGN).unwrap();
+        unsafe {
+            let buffer = alloc::alloc_zeroed(layout);
+            let slice = std::slice::from_raw_parts_mut(buffer, size);
+            slice.copy_from_slice(bytes.as_slice());
+            Self { buffer, size }
+        }
+    }
+
+    pub fn from_string(bytes: String) -> Self {
+        Self::from_vec(bytes.into_bytes())
+    }
+
+    pub fn copy_vec(&mut self, bytes: Vec<u8>) -> &mut Self {
+        self.as_mut_slice().copy_from_slice(bytes.as_slice());
+        self
+    }
+
+    pub fn copy_string(&mut self, bytes: String) -> &mut Self {
+        self.copy_vec(bytes.into_bytes())
+    }
+
+    pub fn as_mut_slice(&mut self) -> &mut [u8] {
+        unsafe { std::slice::from_raw_parts_mut(self.buffer, self.size) }
+    }
+
+    pub fn as_slice(&self) -> &[u8] {
+        unsafe { std::slice::from_raw_parts(self.buffer, self.size) }
+    }
+}
+
+// Interface modelled after std::fs::OpenOptions.
+// It is a builder for the File.
+#[derive(Clone)]
+pub struct OpenOptions {
+    read: bool,
+    write: bool,
+    create: bool,
+}
+
+impl Default for OpenOptions {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl OpenOptions {
+    // Creates a new `OpenOptions` struct with no flag set.
```

Comment:
The comments that you added won't show up as docstrings. Docstrings use three slashes instead of two. If you want to check that your docstrings are rendered properly, you can use `cargo doc` to generate the documentation.